generate.go

```diff
@@ -0,0 +1,203 @@
+	if c.Path == "" || c.Value != "" {
+		return c.Value
+	}
+
+	if c.encrypted {
+		// TODO COGS-1657
+		// decrypt.Open(c.Path, c.SubPath)
+		return "|enc| " + c.Path
+	}
+	// TODO COGS-1659
+	// cogs.Open(c.Path, c.SubPath)
+	return "path." + c.Path + "/" + c.SubPath
+
+}
+
+func (c Cfg) String() string {
+	return fmt.Sprintf(`Cfg{
+	Name: %s
+	Value: %s
+	Path: %s
+	SubPath: %s
+	encrypted: %t
+}`, c.Name, c.Value, c.Path, c.SubPath, c.encrypted)
+}
+
+func (e Env) GenerateMap() map[string]string {
+	cfgMap := make(map[string]string)
+	for k, cfg := range e.Config {
+		cfgMap[k] = cfg.GenerateValue()
+	}
+	return cfgMap
+
+}
+
+// Mapper is an iterface that defines a struct able to generate a flat associative array
+type Mapper interface {
+	GenerateMap() map[string]string
+}
+
+// Generate is a top level command that takes an env argument and cogfilepath to return a string map
+func Generate(env, cogFile string) (map[string]string, error) {
+
+	tree, err := toml.LoadFile(cogFile)
+	if err != nil {
+		return nil, err
+	}
+	return generate(env, tree)
+
+}
+
+func generate(env string, tree *toml.Tree) (map[string]string, error) {
+	var cog Cog
+	var ok bool
+	var err error
+
+	// grab manifest name
+	cog.Name, ok = tree.Get("name").(string)
+	if !ok || cog.Name == "" {
+		return nil, fmt.Errorf("manifest.name string value must be present as a string")
+	}
+
+	var rawManifest map[string]interface{}
+	if err = tree.Unmarshal(&rawManifest); err != nil {
+		return nil, err
+	}
+
+	rawEnv, ok := rawManifest[env]
+	if !ok {
+		return nil, fmt.Errorf("%s environment missing from cog file", env)
+	}
+
+	cog.env.Config, err = ParseEnv(rawEnv)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", env, err)
+	}
+
+	return cog.env.GenerateMap(), nil
+}
+
+// ParseEnv traverses an map interface to return a Cfg string map
+func ParseEnv(rawMap interface{}) (cfgMap map[string]Cfg, err error) {
+	mapEnv, ok := rawMap.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("env must be a table")
+	}
+
+	cfgMap = make(map[string]Cfg)
+
+	// treat enc key as a nested cfgMap
+	if rawEnc, ok := mapEnv["enc"]; ok {
+		encMap, ok := rawEnc.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("enc must be a table")
+		}
+
+		_, err := parseEnv(cfgMap, encMap)
+		if err != nil {
+			return nil, err
+		}
+		for key, cfg := range cfgMap {
+			cfg.encrypted = true
+			cfgMap[key] = cfg
+		}
+		// remove env map now that it is parsed
+		delete(mapEnv, "enc")
+	}
+	// TODO handle same key name declared as env.key_name and env.enc.key_name
+	return parseEnv(cfgMap, mapEnv)
+}
+
+func parseEnv(cfgMap map[string]Cfg, mapEnv map[string]interface{}) (map[string]Cfg, error) {
+	var err error
+	for k, rawCfg := range mapEnv {
+		switch t := rawCfg.(type) {
+		case string:
+			val := rawCfg.(string)
+			cfgMap[k] = Cfg{
+				Name:  k,
+				Value: val,
+			}
+		case map[string]interface{}:
+			val := rawCfg.(map[string]interface{})
```

Comment:
you can actually just use `t` here and dont have to cast like this. Thats what the shorthand `switch t := rawCfg.(type)` does for you, inside the case for the type, `t` is guaranteed to be of that type. Same thing for the `string` case.

Reply:
ah, thanks for that. I assumed it only gave a type signature similar to `reflect.TypeOf`

Reply:
for sure! luckily not, reflect is god awful to work with